### PR TITLE
Handling out of memory for cli-plugin runner (LSF only).

### PR DESCRIPTION
**Pull request by pcm32**
This PR is a first try at handling out of memory for LSF cli plugin. Aims to sort out #6864.

### CODE REVIEW

**Comment by pcm32** [comment]
This is somehow breaking the job status handling (jobs don't turn yellow and then green or red)... however I'm not getting anything useful in the logs... will need to investigate.

**Comment by mvdbeek** [comment]
There's an off chance that https://github.com/galaxyproject/galaxy/pull/6836 might fix monitoring under certain setups. Does your deployment include 02410b270d025e5fb8cb4e2036fb627051c7850a ?

**Comment by pcm32** [comment]
mmm, not sure if it does, but I'm trying it right on top of the previous PR I made for LSF where that was working (even though on this PR I stemmed from up-to-date develop), so I guess it must be something I added....

**Comment by pcm32** [comment]
Everything here seems to be working (so jobs do get ear marked for being out of memory https://github.com/galaxyproject/galaxy/pull/6866/files#diff-4f9c3ae17d7cd09f6e4f353c6ff1ca6eR201) however, the resubmission doesn't seem to be triggered.

I reckon this could be an issue of my local setup in that I'm using dynamic destinations. Would this construct work with dynamic destinations (in my job_conf.xml)?

```
    <destination id="dynamic-lsf-micro" runner="dynamic">
      <param id="type">python</param>
      <param id="function">lsf_wrapper_micro</param>
      <resubmit condition="memory_limit_reached" destination="dynamic-lsf-tiny"/>
    </destination>
    <destination id="dynamic-lsf-tiny" runner="dynamic">
      <param id="type">python</param>
      <param id="function">lsf_wrapper_tiny</param>
      <resubmit condition="memory_limit_reached" destination="dynamic-lsf-small"/>
    </destination>
```

or my python code in that dynamic destination would need to do something so that this gets triggered? Thanks!

**Comment by pcm32** [comment]
mmm... no, lack of resubmission seems unrelated to the usage of dynamic destinations, I just tried with non dynamic destinations:

```
    <destinations default="local_lsf_1cpu_500MB_Ram">
    <!-- <destinations default="dynamic-lsf-small"> -->
        <destination id="local" runner="local"/>

        <!-- Example CLI LSF Runner: 8 cores and 16 GB ram -->
        <destination id="local_lsf_8cpu_16GbRam" runner="cli">
            <param id="shell_plugin">LocalShell</param>
            <param id="job_plugin">LSF</param>
            <param id="job_memory">16000</param>
            <param id="job_cores">7</param>
        </destination>

        <destination id="local_lsf_1cpu_500MB_Ram" runner="cli">
            <param id="shell_plugin">LocalShell</param>
            <param id="job_plugin">LSF</param>
            <param id="job_memory">500</param>
            <param id="job_cores">1</param>
            <resubmit condition="memory_limit_reached" destination="local_lsf_8cpu_16GbRam"/>
        </destination>
 ....
</destinations>
```

and the result is the same... no re-submission.. :-(

**Comment by pcm32** [comment]
Apparently the lack of resubmission was due to the cli plugin not ever marking failed jobs as such. Maybe it is done at another level that I couldn't find, but adding the `self.mark_as_failed(ajs)` as in my last commit effectively triggers the resubmission when the resubmission condition is fulfilled.

**Comment by pcm32** [comment]
The circle-ci error seems to be unrelated to the content of the PR, all the rest is passing:

```
=================================== FAILURES ===================================
_______________ ToolBoxTestCase.test_tool_reload_for_broken_tool _______________

self = <unit.tools.test_toolbox.ToolBoxTestCase testMethod=test_tool_reload_for_broken_tool>

    def test_tool_reload_for_broken_tool(self):
        self._init_tool(filename="simple_tool.xml", version="1.0")
        self._add_config("""<toolbox><tool file="simple_tool.xml"/></toolbox>""")
        toolbox = self.toolbox
        tool = toolbox.get_tool('test_tool')
        assert tool is not None
        assert tool.version == "1.0"
        assert tool.tool_errors is None
        # Tool is loaded, now let's break it
        tool_path = tool.config_file
        with open(tool.config_file, 'w') as out:
            out.write('certainly not a valid tool')
    
        def check_tool_errors():
            tool = self.toolbox.get_tool("test_tool")
            assert tool is not None
            assert tool.version == "1.0"
            assert tool.tool_errors == 'Current on-disk tool is not valid'
    
>       self._try_until_no_errors(check_tool_errors)

test/unit/tools/test_toolbox.py:225: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/unit/tools/test_toolbox.py:249: in _try_until_no_errors
    raise e
test/unit/tools/test_toolbox.py:243: in _try_until_no_errors
    f()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def check_tool_errors():
        tool = self.toolbox.get_tool("test_tool")
        assert tool is not None
        assert tool.version == "1.0"
>       assert tool.tool_errors == 'Current on-disk tool is not valid'
E       AssertionError: assert None == 'Current on-disk tool is not valid'
E         -None
E         +'Current on-disk tool is not valid'

test/unit/tools/test_toolbox.py:223: AssertionError
----------------------------- Captured stderr call -----------------------------
INFO:galaxy.tools.toolbox.base:Parsing the tool configuration /tmp/tmpr0ps4gif/tool_conf.xml
DEBUG:galaxy.tools.toolbox.base:Loaded tool id: test_tool, version: 1.0 into tool panel..
DEBUG:galaxy.tools.toolbox.base:Reading tools from config files finished (1.134 ms)
------------------------------ Captured log call -------------------------------
base.py                    158 INFO     Parsing the tool configuration /tmp/tmpr0ps4gif/tool_conf.xml
base.py                    341 DEBUG    Loaded tool id: test_tool, version: 1.0 into tool panel..
base.py                    137 DEBUG    Reading tools from config files finished (1.155 ms)
=============================== warnings summary ===============================
.tox/py35-unit/lib/python3.5/site-packages/bioblend/config.py:22
  /home/circleci/repo/.tox/py35-unit/lib/python3.5/site-packages/bioblend/config.py:22: DeprecationWarning: The SafeConfigParser class has been renamed to ConfigParser in Python 3.2. This alias will be removed in future versions. Use ConfigParser directly instead.
    'debug': '0'})

.tox/py35-unit/lib/python3.5/site-packages/sqlalchemy/orm/relationships.py:2624
  /home/circleci/repo/.tox/py35-unit/lib/python3.5/site-packages/sqlalchemy/orm/relationships.py:2624: SAWarning: Non-simple column elements in primary join condition for property LibraryDatasetDatasetAssociation.visible_children - consider using remote() annotations to mark the remote side.
    "remote() annotations to mark the remote side." % self.prop

.tox/py35-unit/lib/python3.5/site-packages/sqlalchemy/orm/relationships.py:2624
  /home/circleci/repo/.tox/py35-unit/lib/python3.5/site-packages/sqlalchemy/orm/relationships.py:2624: SAWarning: Non-simple column elements in primary join condition for property HistoryDatasetAssociation.visible_children - consider using remote() annotations to mark the remote side.
    "remote() annotations to mark the remote side." % self.prop

lib/galaxy/model/orm/scripts.py::galaxy.model.orm.scripts.get_config
lib/galaxy/model/orm/scripts.py::galaxy.model.orm.scripts.get_config
  /home/circleci/repo/lib/galaxy/model/orm/scripts.py:2: DeprecationWarning: The SafeConfigParser class has been renamed to ConfigParser in Python 3.2. This alias will be removed in future versions. Use ConfigParser directly instead.
    Code to support database helper scripts (create_db.py, manage_db.py, etc...).
```

I have tested the functionality (through the LSF cli runner) and all is working well.

**Comment by pcm32** [comment]
For the reviewers, the only controversial part of this PR could be:

https://github.com/galaxyproject/galaxy/pull/6866/files#diff-4f9c3ae17d7cd09f6e4f353c6ff1ca6eR183

But without that, resubmission doesn't get triggered.

**Comment by mvdbeek** [comment]
Thanks @pcm32, I will look into this.

> Maybe it is done at another level that I couldn't find, but adding the `self.mark_as_failed(ajs)` as in my last commit effectively triggers the resubmission when the resubmission condition is fulfilled.

Yes, it is triggered in the `finish_job` section common to all runners https://github.com/galaxyproject/galaxy/blob/0fe13b7e5af60b1bf7417a1a5ff981dd21cb026f/lib/galaxy/jobs/runners/__init__.py#L765, just before calling`_finish_or_resubmit_job`.
But this will only work for error messages coming directly from the tool.
The only runner with working resubmission on actual job errors seems to be slurm, and slurm
does (almost )the same in https://github.com/galaxyproject/galaxy/blob/9a26de1b4b5a96eacab97aca2e74b994f38bdc31/lib/galaxy/jobs/runners/slurm.py#L160
so this looks OK to me in principle.

File: lib/galaxy/jobs/runners/cli.py [hunk 1]
```diff
@@ -177,6 +177,10 @@ def check_watched_items(self):
                 if not state == model.Job.states.OK:
                     # No need to change_state when the state is OK, this will be handled by `self.finish_job`
                     ajs.job_wrapper.change_state(state)
+                if state == model.Job.states.ERROR:
+                    # Try to find out the reason for exiting
+                    self.__handle_out_of_memory(ajs, external_job_id)
+                    self.mark_as_failed(ajs)
```
Comment (mvdbeek) [inline lib/galaxy/jobs/runners/cli.py]:
```suggestion
                    self.work_queue.put((self.mark_as_failed, ajs))
```

Comment (pcm32) [reply lib/galaxy/jobs/runners/cli.py]: Thanks, happy to go with this if you say it is more correct. I will test it locally to make sure it works and then commit the suggestion.

**Comment by pcm32** [comment]
I have checked that the way you proposed @mvdbeek for setting the job as failed works fine. All working for me on this PR.

**Review by mvdbeek — APPROVED** [review]
Great stuff, thanks @pcm32! I should really do the same for the torque plugin.

**Comment by mvdbeek** [comment]
@galaxybot test py3

**Pull request merged**